### PR TITLE
Precheck target controller migrations

### PR DESCRIPTION
**Pull request by SimonRichardson**
Ensure we pass all the facades that are available to the client so that it can verify with the target controller that there is an overlap. This is to ensure that the agent doesn't become wedged if certain API facades aren't available. The source client passes all the facades available to it. It's far better to be conservative with what we send, by sending all of the facades that a subset. The target controller pre-check only checks for a subset. We could add more facades, but we can discuss that in the PR. The facade `CompleteIntersection` code only checks for a partial overlap. If a facade called `Foo`, with versions `1,2,3` on the client is sent to the target controller, and the target controller has version `3,4,5`. We have an intersection of `3`. The migration can continue. This allows us to remove old facades, as long as there is some sort of overlap.

This should still work against 3.x controllers, as we only validate against all facades if the version 3 or greater. Ideally, we'll move all facades up to 3 once that's landed. Allowing us to be a lot more secure.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd src --build-agent
$ juju add-model default
$ juju deploy ubuntu
```

```sh
$ juju bootstrap lxd dst --build-agent
$ juju switch src
```

Watching both controllers in two different terminals:

```sh
$ watch juju status -m src:default
$ watch juju status -m dst:default
```
The latter one won't appear until the model is migrated.

```sh
$ juju migrate src:default dst
```

## Links

**Jira card:** JUJU-4962

**LP Bug:** https://bugs.launchpad.net/juju/+bug/2043986

### CODE REVIEW

File: api/facadeversions.go [hunk 1]
```diff
@@ -135,14 +140,3 @@ var facadeVersions = map[string][]int{
 	"VolumeAttachmentsWatcher":     {2},
 	"VolumeAttachmentPlansWatcher": {1},
 }
-
-// bestVersion tries to find the newest version in the version list that we can
-// use.
-func bestVersion(desired []int, versions []int) int {
```
Comment (SimonRichardson) [inline api/facadeversions.go:141]: Move to core package to become more useful.

**Review by manadart — COMMENTED** [review]
Looks good, and QA passes. Just one question over the message and/or whether we block major version leap-frog.

**Review by manadart — APPROVED** [review]
Thanks for the changes.

File: apiserver/allfacades.go [hunk 2]
```diff
@@ -107,8 +106,73 @@ import (
...
+func requiredMigrationFacadeVersions() facades.FacadeVersions {
+	registry := new(facade.Registry)
+
+	// Client and modelmanager facades are required for the migration
+	// master to function correctly. Missing a model manager causes the
+	// status to error out.
+	client.Register(registry)
+	modelmanager.Register(registry)
+
+	// The following are required to keep the agent alive during
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/allfacades.go:128]: Added a comment about this being ultra conservative

**Pull request merged**